pkg/cache/cache.go

```diff
@@ -0,0 +1,260 @@
+		workqueue:       workqueue.NewRateLimitingQueue(workqueue.DefaultControllerRateLimiter()),
+		ListFunc:        args.ListFunc,
+		ObjectType:      args.ObjectType,
+		log:             log.WithFields(log.Fields{"type": args.ObjectType}),
+		mut:             &sync.Mutex{},
+	}
+}
+
+func (c *calicoCache) Set(key string, newObj interface{}) {
+	if reflect.TypeOf(newObj) != c.ObjectType {
+		c.log.Fatalf("Wrong object type recieved to store in cache. Expected: %s, Found: %s", c.ObjectType, reflect.TypeOf(newObj))
+	}
+
+	// Check if the object exists in the cache already.  If it does and hasn't changed,
+	// then we don't need to send an update on the queue.
+	if existingObj, found := c.threadSafeCache.Get(key); found {
+		c.log.Debugf("%#v already exists in cache - comparing.", existingObj)
+		if !reflect.DeepEqual(existingObj, newObj) {
+			// The objects do not match - send an update over the queue.
+			c.threadSafeCache.Set(key, newObj, cache.NoExpiration)
+			if c.isRunning() {
+				c.log.Debugf("Queueing update - %#v and %#v do not match.", newObj, existingObj)
+				c.workqueue.Add(key)
+			}
+		}
+	} else {
+		c.threadSafeCache.Set(key, newObj, cache.NoExpiration)
+		if c.isRunning() {
+			c.log.Debugf("%#v not found in cache, adding it + queuing update.", newObj)
+			c.workqueue.Add(key)
+		}
+	}
+}
+
+func (c *calicoCache) Delete(key string) {
+	c.log.Debugf("Deleting %s from cache", key)
+	c.threadSafeCache.Delete(key)
+	c.workqueue.Add(key)
+}
+
+func (c *calicoCache) Clean(key string) {
+	c.log.Debugf("Cleaning %s from cache, no update required", key)
+	c.threadSafeCache.Delete(key)
+}
+
+func (c *calicoCache) Get(key string) (interface{}, bool) {
+	obj, found := c.threadSafeCache.Get(key)
+	if found {
+		return obj, true
+	}
+	return nil, false
+}
+
+// Prime adds the key and value to the cache but will never generate
+// an update on the queue.
+func (c *calicoCache) Prime(key string, value interface{}) {
+	c.threadSafeCache.Set(key, value, cache.NoExpiration)
+}
+
+// ListKeys returns a list of all the keys in the cache.
+func (c *calicoCache) ListKeys() []string {
+	cacheItems := c.threadSafeCache.Items()
+	keys := make([]string, 0, len(cacheItems))
+	for k := range cacheItems {
+		keys = append(keys, k)
+	}
+
+	return keys
+}
+
+// GetQueue returns the output queue from the cache.  Whenever a key/value pair
+// is modified, an event will appear on this queue.
+func (c *calicoCache) GetQueue() workqueue.RateLimitingInterface {
+	return c.workqueue
+}
+
+// Run starts the cache.  Any Set() calls prior to calling Run() will
+// prime the cache, but not trigger any updates on the output queue.
+func (c *calicoCache) Run(reconcilerPeriod string) {
+	go c.reconcile(reconcilerPeriod)
+
+	// Indicate that the cache is running, and so updates
+	// can be queued.
+	c.mut.Lock()
+	c.running = true
+	c.mut.Unlock()
+}
+
+func (c *calicoCache) isRunning() bool {
+	c.mut.Lock()
```

Comment:
do we really need a mutex lock for reading?

Reply:
Probably not tbh, but I was rushing and was being safe :)